diff --git a/restclient.module b/restclient.module
index 12108b9..c3b9a82 100644
--- a/restclient.module
+++ b/restclient.module
@@ -69,6 +69,55 @@ function restclient_wsconfig_connector_info() {
   );
 }
 
+
+/**
+ * Implements hook_rules_event_info().
+ */
+function restclient_rules_event_info() {
+  return array(
+    'rest_get_failure' => array(
+      'label' => t('Rest call failed'),
+      'group' => t('Rest'),
+      'variables' => array(
+        'response' => array(
+          'type' => 'text',
+          'label' => t('Response number.')
+        ),
+        'rest_callback' => array(
+          'type' => 'text',
+          'label' => t('The Rest call being called.')
+        ),
+      ),
+    ),
+  );
+}
+
+
+/**
+ * Implements hook_rules_action_info().
+ */
+/**
+function restclient_rules_action_info() {
+  return array(
+    'mail_rest_errors' => array(
+      'group' => t('System'),
+      'label' => t('Mail rest errors'),
+      'parameter' => array(
+        'response' => array(
+          'type' => 'text',
+          'label' => t('Response number.')
+        ),
+        'rest_callback' => array(
+          'type' => 'text',
+          'label' => t('The Rest call being called.')
+        ),
+      ),
+    ),
+  );
+}
+*/
+
+
 /**
  * Make a GET request
  *
@@ -163,10 +212,12 @@ function restclient_get($resource_path, $variables = array()) {
         cache_set($cid, $response, 'cache_restclient', $expires);
       }
     }
-
     // No error occured, return the response
     return $response;
   }
+  rules_invoke_event('rest_get_failure', $response->code, $url);
+
+
 
   // Log the error and return FALSE
   _restclient_watchdog('GET', $response->code,  $response->error, $url);
