diff --git a/includes/content.inc b/includes/content.inc
index 890ec23..d2e719a 100644
--- a/includes/content.inc
+++ b/includes/content.inc
@@ -298,7 +298,7 @@ function ctools_content_render($type, $subtype, $conf, $keywords = array(), $arg
       // Give previous title as an available substitution here.
       $keywords['%title'] = empty($content->title) ? '' : $content->title;
       $content->original_title = $keywords['%title'];
-      $content->title = $conf['override_title_text'];
+      $content->title = t($conf['override_title_text']);
     }
 
     if (!empty($content->title)) {
